### PR TITLE
Perform actions when aqua scan detects security vulnerabilities with solutions

### PR DESCRIPTION
**Pull request by valituguran**
Actions performed:

- Failing a pipeline when a security vulnerability with a solution is found
- Deleting the vulnerable image
- Checking that the scan runs correctly
- Add a blocking entry to the code insight sent to Bitbucket

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # Changelog
 
 ## Unreleased
+* Perform actions when aqua scan detects security vulnerabilities with solutions ([#1147](https://github.com/opendevstack/ods-jenkins-shared-library/pull/1147))
```
Comment (tbugfinder) [inline CHANGELOG.md:4]: Actually the code is more specific. It's about remotely exploitable vulnerabilities, not about any possible fix.
Please be precise for a changelog reader.

**Comment by tbugfinder** [comment]
Doc should also be updated.
https://github.com/opendevstack/ods-jenkins-shared-library/blob/master/docs/modules/jenkins-shared-library/partials/odsComponentStageScanWithAqua.adoc

File: src/org/ods/component/ScanWithAquaStage.groovy [hunk 2]
```diff
@@ -283,7 +314,27 @@ class ScanWithAquaStage extends Stage {
...
+    private void setCurrentBuildUnstable() {
+        this.steps.currentBuild.result = 'UNSTABLE'
+    }
+
+    private List filterRemoteCriticalWithSolutionVulnerabilities(Map aquaJsonMap) {
+        List result = []
+        aquaJsonMap.resources.each { it ->
+            (it as Map).vulnerabilities.each { vul ->
+                Map vulnerability = vul as Map
+                if ((vulnerability?.exploit_type as String)?.equalsIgnoreCase("remote")
```
Comment (hrcornejo) [inline src/org/ods/component/ScanWithAquaStage.groovy:331]: Use constant for "remote" and "critical"

File: src/org/ods/component/ScanWithAquaStage.groovy [hunk 3]
```diff
@@ -199,13 +215,28 @@ class ScanWithAquaStage extends Stage {
                 [ title: "Messages", value: prepareMessageToBitbucket(messages), ]
             ])
         }
+        if (actionableVulnerabilities?.size() > 0) {
```
Comment (hrcornejo) [inline src/org/ods/component/ScanWithAquaStage.groovy:218]: Why not add a new message per actionableVulnerabilitie like: title: "Explotable vulnerability", value: Information about the vulnerability (using prepareMessageToBitbucket)

File: src/org/ods/component/ScanWithAquaStage.groovy [hunk 4]
```diff
@@ -199,13 +215,28 @@ class ScanWithAquaStage extends Stage {
...
+            if (data.messages) {
+                ((List) data.messages).add([
+                    title: "Blocking",
+                    value: "Yes"
+                ])
+            } else {
+                data.put("messages", [
```
Comment (hrcornejo) [inline src/org/ods/component/ScanWithAquaStage.groovy:225]: Try to avoid duplicate parts. If "messages" is needed, add it and later add the information

**Review by hrcornejo — CHANGES_REQUESTED** [review]

File: src/org/ods/component/ScanWithAquaStage.groovy [hunk 5]
```diff
@@ -199,13 +215,28 @@ class ScanWithAquaStage extends Stage {
...
+            }
+        }
 
         bitbucket.createCodeInsightReport(data, context.repoName, context.gitCommit)
     }
 
     private createBitbucketCodeInsightReport(String messages) {
         String title = "Aqua Security"
-        String details = "There was some problems with Aqua:"
+        String details = "There were some problems with Aqua"
```
Comment (hrcornejo) [inline src/org/ods/component/ScanWithAquaStage.groovy:239]: Maintain the ":" at the end:
image

File: src/org/ods/component/ScanWithAquaStage.groovy (hunk 3 above)
Comment (valituguran) [reply src/org/ods/component/ScanWithAquaStage.groovy:218]: We need this specific entry to be present in the data structure as a marker: {
            "title": "Blocking",
            "type": "TEXT",
            "value": "Yes"
        }

**Comment by hrcornejo** [comment]
Release Manager with a component containing security vulnerabilities with a solution:
- Failing a pipeline when a security vulnerability with a solution is found
image
image
- Deleting the vulnerable image
image
Non existing tag:
image
- Add a blocking entry to the code insight sent to Bitbucket
image
- Checking that the scan runs correctly
image

Single component containing security vulnerabilities with a solution:
- Also failing the pipeline, deleting image:
image

**Review by hrcornejo — APPROVED** [review]

**Pull request merged**